api/base/serializers.py

```diff
@@ -80,6 +81,18 @@ def to_internal_value(self, data):
         return super(TypeField, self).to_internal_value(data)
 
 
+class TargetTypeField(TypeField):
+    """
+    Enforces that the related resource has the correct type
+    """
+
+    def to_internal_value(self, data):
+        if self.root.Meta.target_type_ != data:
+            raise Conflict()
+        # Super call intentional to prevent target_type_ being checked against type_
+        return super(TypeField, self).to_internal_value(data)
```

Comment:
If I did super(TargetTypeField) it would try to check target_type (users) against the type_ (contributors), for example.